crates/ruff_linter/src/suppression.rs

```diff
@@ -0,0 +1,905 @@
+use core::fmt;
+use ruff_python_parser::{TokenKind, Tokens};
+use ruff_source_file::LineRanges;
+use std::{error::Error, fmt::Formatter};
+use thiserror::Error;
+
+use ruff_python_trivia::{Cursor, is_python_whitespace};
+use ruff_text_size::{Ranged, TextLen, TextRange, TextSize, TextSlice};
+use smallvec::{SmallVec, smallvec};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum SuppressionAction {
+    Disable,
+    Enable,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct SuppressionComment {
+    /// Range containing the entire suppression comment
+    range: TextRange,
+
+    /// How indented an own-line comment is, or None for trailing comments
+    indent: Option<String>,
+
+    /// The action directive
+    action: SuppressionAction,
+
+    /// Ranges containing the lint codes being suppressed
+    codes: SmallVec<[TextRange; 2]>,
+
+    /// Range containing the reason for the suppression
+    reason: TextRange,
+}
+
+impl SuppressionComment {
+    /// Return the suppressed codes as strings
+    fn codes_as_str(&self, source: &str) -> Vec<String> {
+        self.codes
+            .iter()
+            .map(|range| source.slice(range).to_string())
+            .collect()
+    }
+
+    /// Whether the comment "matches" another comment, based on indentation and suppressed codes
+    fn matches(&self, other: &SuppressionComment, source: &str) -> bool {
+        ((self.action == SuppressionAction::Enable && other.action == SuppressionAction::Disable)
+            || (self.action == SuppressionAction::Disable
+                && other.action == SuppressionAction::Enable))
+            && self.indent == other.indent
+            && self.codes_as_str(source) == other.codes_as_str(source)
+    }
+}
+
+#[allow(unused)]
+#[derive(Clone, Debug)]
+pub(crate) struct Suppression {
+    /// The lint code being suppressed
+    code: String,
+
+    /// Range for which the suppression applies
+    range: TextRange,
+
+    /// Any comments associated with the suppression
+    comments: SmallVec<[SuppressionComment; 2]>,
```

Comment:
`SmallVec` stack-allocates up to `N` items before heap allocating a regular `Vec`. This has the advantage that it reduces the pressure on heap size. However, it can have the downside that the containing type (`Suppression`) becomes larger because the `SmallVec` requires at least `N * SuppressionComment` bytes to store the inline elements. 

The main downside of small vec is that it wastes memory when the stored elements is less than `N`. It also makes copying `Suppression` more expensive because it requires writing more bytes (e.g. when adding a `Suppression` to a `Vec`)


Given that `SuppressionComment` is a fairly chunky type (72 bytes) (rust-analyzer shows you the size on hover) and the most common case is probably to suppress exactly one code, I suggest reducing `N` to 1
```suggestion
    comments: SmallVec<[SuppressionComment; 1]>,
```

Reply:
> Given that SuppressionComment is a fairly chunky type (72 bytes) (rust-analyzer shows you the size on hover) and the most common case is probably to suppress exactly one code, I suggest reducing N to 1

Also I think you misunderstand the current use here. In the expected case, this vec would contain a reference to both the disable and enable `SuppressionComment` objects. Unmatched disable/enable would end up in the `invalid` pile rather than getting a `Suppression` object created.

Reply:
I'm leaning towards making this a lint rule, which we can enable by default, but giving users a way to opt out if they want to.